### PR TITLE
feat(webdriver-manager): Configurable output directory

### PR DESCRIPTION
**Pull request by blabno**
Allows to change binaries cache directory, so that you don't need to download them each time you remove node_modules and run npm install and webdriver-manager update.

### CODE REVIEW

**Comment by blabno** [comment]
I signed it!

**Comment by sjelin** [comment]
Why an env variable instead of a flag?

**Comment by blabno** [comment]
For the same reason as http_proxy, so that we can setup env and keep build
descriptors env agnostic.
23 wrz 2015 19:53 "Sammy Jelin" <notifications@github.com> napisał(a):

**Comment by blabno** [comment]
Particularly I have platform where I run tasks, mini projects, completed by
other people remotely on their PCs. Most of the dependencies are the same
so I want my Jenkins to reuse dependencies between build to save bandwidth.

If I was to put path to out-dir in package.jso n then it would be very
tightly coupled to my env, and may conflict with other people config.

Besides there's already a flag. This PR adds additional flexibility.
23 wrz 2015 20:00 "Bernard Łabno" <s4237@pjwstk.edu.pl> napisał(a):

File: bin/webdriver-manager [hunk 1]
```diff
@@ -100,7 +100,7 @@ var cli = optimist.
         '  update: install or update selected binaries\n' +
         '  start: start up the selenium server\n' +
         '  status: list the current available drivers').
-    describe('out_dir', 'Location to output/expect ').
+    describe('out_dir', 'Location to output/expect binaries, can be modified with WEBDRIVER_MANAGER_SELENIUM_DIR evn variable').
```
Comment (juliemr) [inline bin/webdriver-manager]: typo: evn/env

**Comment by juliemr** [comment]
My gut reaction is that there must be a better way to do this - why do you need to change your `package.json` to couple a command line flag?

The reason we support HTTP_PROXY as an environment variable is because this is an existing convention that users expect to work for all node projects on their machine. I'm wary of introducing new environment variables.

**Comment by blabno** [comment]
Ok, lets imagine that my package.json looks like this:
```
"scripts": {
  "pretest":"webdriver-manager update --standalone --out_dir=/home/bernard/.protractor-cache"
}
```

Now Mikey, won't be too happy because he would like to store the binaries in `/home/mikey`.
Env variable solves this nicely.

**Comment by blabno** [comment]
btw. the reason why I put `webdriver-manager update` into package.json is because I want to limit number of commands my students have to type before running tests to minimum.

Of course I could write for each task that I give them:
install npm
install grunt
install bower
install protractor
update webdriver
etc.

But isn't this nicer `npm install && npm test`?

**Comment by juliemr** [comment]
But, you still have to tell them to set the environment variable! 

I think the better way to solve this is:

```
  "config": {
    "webdriver_manager_selenium_dir": "/abc/def/ghi"
  },
  "scripts": {
    "pretest":"webdriver-manager update --standalone --out_dir=$npm_package_config_webdriver_manager_selenium_dir"
  }
```

**Comment by juliemr** [comment]
I'm closing this as I still don't think we should be adding new, non-standard environment variables, and I'm convinced that npm already provides the tools to do this. Thanks for your work (and for using Protractor in your teaching!).

**Comment by blabno** [comment]
Pity. I'll have to force users to use my fork.
10 lis 2015 22:31 "Julie Ralph" <notifications@github.com> napisał(a):

**Pull request closed without merging**